relay-server/src/actors/spooler/mod.rs

```diff
@@ -501,18 +394,330 @@ impl BufferService {
+            if !disk.is_empty().await.unwrap_or_default() {
+                state =
+                    BufferState::to_memory_file_ro(config.spool_envelopes_max_memory_size(), disk)
+            }
+        }
+
+        Ok(Self {
+            services,
+            state,
+            config,
+        })
+    }
+
+    /// Handles the enqueueing messages into the internal buffer.
+    async fn handle_enqueue(&mut self, message: Enqueue) -> Result<(), BufferError> {
+        let Enqueue {
+            key,
+            value: managed_envelope,
+        } = message;
 
-        // Refresh DB state only if we get the proper reading on the file.
-        if let Ok(estimated_size) = Self::estimate_buffer_size(db).await {
-            *is_disk_full = (estimated_size as usize) >= *max_disk_size;
+        match self.state {
+            BufferState::Memory { ref mut ram } => ram.enqueue(key, managed_envelope),
+            // Will buffer till either there free space on disk, and the in-memory buffer can be
+            // spooled, or there is OOM.
+            // NOTE: Right now the number of envelopes is still restricted by
+            // `cache.envelope_buffer_size` and they will be dropped once the limit is reached.
+            BufferState::MemoryFileRead {
+                ref mut ram,
+                ref disk,
+            } => {
+                ram.enqueue(key, managed_envelope);
+
+                // If there is enough space on the disk for the in-memory buffer, spool
+                // everyhting to disk and switch state to `BufferState::File`.
+                if let Ok(spool_size) = disk.estimate_spool_size().await {
+                    if disk
+                        .max_disk_size
+                        .checked_sub(spool_size as usize)
+                        .map_or(false, |s| s > ram.used_memory)
+                    {
+                        disk.spool(&mut ram.buffer, &mut ram.used_memory).await?;
+                        self.state = BufferState::to_file(disk);
+                    }
+                }
+            }
+            BufferState::MemoryFileStandby {
+                ref mut ram,
+                ref disk,
+            } => {
+                ram.enqueue(key, managed_envelope);
+
+                // If memory buffer is full, spool to disk and switch state to
+                // `BufferState::File`.
+                if ram.is_full() {
+                    disk.spool(&mut ram.buffer, &mut ram.used_memory).await?;
+                    self.state = BufferState::to_file(disk);
+                }
+            }
+            BufferState::File { ref disk } => {
+                disk.enqueue(key, managed_envelope).await?;
+
+                // The disk is full, switch state to `BufferState::Memory` and try to buffer as
+                // much as we can before hitting OOM.
+                if disk.is_full().await.unwrap_or_default() {
+                    self.state = BufferState::to_memory_file_ro(
```

Comment:
I'm questioning whether we actually need the `MemoryFileRead` state. This state is so extreme that we might as well just start dropping events. That would simplify the state machine a lot. But I might be missing something.

Reply:
So we want to make sure we spool to disk as much as we can, and then also use memory as much as we can before start rejecting incoming events.  `MemoryFileRead` signals us that the disk is full and can only be read from, but we still can try to fill up the memory, which right now is guarded by `cache.envelope_buffer_size`, but it will get deprecated and we will actually use the new config from `spool.envelopes.max_memory_size`. 
This is basically also a preparation for it. 
Does it makes sense or maybe I think about this somewhat wrong?